js/views/conversation_view.js

```diff
@@ -523,18 +534,16 @@
             var collection = this.model.messageCollection;
             var length = collection.length;
             var viewportBottom = this.view.outerHeight;
-            var unreadCount = this.model.get('unreadCount');
-
-            if (!unreadCount || unreadCount < 1) {
-                return;
-            }
+            var unreadCount = this.model.get('unreadCount') || 0;
 
             // Start with the most recent message, search backwards in time
             var foundUnread = 0;
             for (var i = length - 1; i >= 0; i -= 1) {
-                // We don't want to search through all messages, so we stop after we've
-                //   hit all unread messages. The unread should be relatively recent.
-                if (foundUnread >= unreadCount) {
+                // Search the latest 30, then stop if we believe we've covered all known
+                //   unread messages. The unread should be relatively recent.
+                // Why? local notifications can be unread but won't be reflected the
+                //   conversation's unread count.
+                if (i > 30 && foundUnread >= unreadCount) {
```

Comment:
30 is certainly arbitrary. About 1/3 of what we pull by default? A couple/several screens' worth? Really what it does is allow for some amount of searching when our conversation `unreadCount` is zero.

Certainly for the 'initial load of conversation' scenario I think a database index would really help us - we could easily find out when the oldest message was. We could potentially then continue to use that value, maybe even a reference to the oldest message, over time. It would help in all sorts of scans, like when we choose where to put the last seen indicator.

For this algorithm in particular, we're starting at the bottom and looking for unread messages whose bottom are above the bottom of the viewport. One thing which might help would be checking (maybe not even every message, but some multiple) to see if the bottom of the current message is above the top of the viewport. That's another sign we could stop. I had been resisting that up to this point, because doing those height calculations for every message would be costly. Using `unreadCount` was my answer there. But now that we can no longer rely on it, maybe it's worth making a change like that...